Mathlib/Computability/NFA.lean

```diff
@@ -112,6 +120,275 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+end Auxilary
+
+/-- NFAs are closed under reversal:
+  Given NFA `M`, there is an NFA `reverse(M)` such that
+  `L(reverse(M)) = reverse(L(M))`. -/
+def reverse (M : NFA α σ) : NFA α σ where
+  step := M.unstep
+  start := M.accept
+  accept := M.start
+
+lemma reverse_acceptsFrom_rewindsToStart {M : NFA α σ} :
+    M.reverse.acceptsFrom = M.rewindsToStart := by
+  ext xs
+  rfl
+
+lemma reverse_rewindFrom_evalFrom {xs : List α} {S1 S2 : Set σ} {M : NFA α σ} :
+    (∃ s1 ∈ S1, s1 ∈ M.rewindFrom S2 xs) ↔
+    (∃ s2 ∈ S2, s2 ∈ M.evalFrom S1 xs.reverse) := by
+  dsimp [evalFrom, rewindFrom]
+  rw [List.foldl_reverse]
+  revert S1 S2
+  induction xs <;> intros S1 S2 <;> simp
+  case nil =>
+    constructor <;> rintro ⟨s, h1, h2⟩ <;> exists s
+  case cons x xs ih =>
+    rw [ih]
+    clear ih
+    unfold unstepSet
+    unfold stepSet
+    constructor
+    · rintro ⟨s3, h, h1⟩
+      simp [Set.mem_iUnion] at h
+      rcases h with ⟨s2,h2,h3⟩
+      exists s2
+      constructor <;> try assumption
+      simp [Set.mem_iUnion]
+      exists s3
+    · rintro ⟨s2, h2, h⟩
+      simp [Set.mem_iUnion] at h
+      rcases h with ⟨s3, h1, h3⟩
+      exists s3
+      simp [Set.mem_iUnion]
+      constructor <;> try assumption
+      exists s2
+
+lemma reverse_rewindsToStart_acceptsFrom {xs : List α} {M : NFA α σ} :
+    xs ∈ M.rewindsToStart M.reverse.start ↔ xs.reverse ∈ M.acceptsFrom M.start := by
+  rw [mem_rewindsToStart, mem_acceptsFrom]
+  apply reverse_rewindFrom_evalFrom
+
+theorem reverse_accepts {M : NFA α σ} :
+    M.reverse.accepts = { xs : List α | xs.reverse ∈ M.accepts } := by
+  ext xs
+  rw [accepts_acceptsFrom, accepts_acceptsFrom, reverse_acceptsFrom_rewindsToStart, Set.mem_setOf]
+  apply reverse_rewindsToStart_acceptsFrom
+
+end Reversal
+
+section Union
+
+variable {σ1 : Type v} {σ2 : Type v}
+
+private instance Language.instUnion : Union (Language α) := by
+  apply Set.instUnion
+
+/--`stepSum M₁ M₂` computes the transition for `M₁ ∪ M₂`. -/
+def stepSum (M1 : NFA α σ1) (M2 : NFA α σ2)
+  (s : σ1 ⊕ σ2) (a : α) : Set (σ1 ⊕ σ2) :=
+    { s' : σ1 ⊕ σ2 |
+      Sum.LiftRel
+        (fun s1 s1' ↦ s1' ∈ M1.step s1 a)
+        (fun s2 s2' ↦ s2' ∈ M2.step s2 a)
+        s s' }
+
+/-- NFAs are closed under union:
+  Given NFAs `M₁` and `M₂`, `M₁ ∪ M₂` is a NFA such that
+  `L(M₁ ∪ M₂) = L(M₁) ∪ L(M₂)`. -/
+def union (M1 : NFA α σ1) (M2 : NFA α σ2) : NFA α (σ1 ⊕ σ2) where
+  start : Set (σ1 ⊕ σ2) := { s : σ1 ⊕ σ2 | s.casesOn M1.start M2.start }
+  step : σ1 ⊕ σ2 → α → Set (σ1 ⊕ σ2) := stepSum M1 M2
+  accept : Set (σ1 ⊕ σ2) := { s : σ1 ⊕ σ2 | s.casesOn M1.accept M2.accept }
+
+lemma union_biUnion_spec
+  {x : α} {S1 : Set σ1} {S2 : Set σ2} {M1 : NFA α σ1} {M2 : NFA α σ2} :
+    (⋃ s, ⋃ (_ : Sum.rec S1 S2 s), stepSum M1 M2 s x)
+    =
+    {s' | Sum.rec
+      (fun s1' ↦ (⋃ (s1 : σ1) (_ : s1 ∈ S1), M1.step s1 x) s1')
+      (fun s2' ↦ (⋃ (s2 : σ2) (_ : s2 ∈ S2), M2.step s2 x) s2') s'} := by
```

Comment:
You're still treating sets as functions here.

Reply:
Ah, sorry, thanks for catching this! I changed the lemma statement to what you suggested in the other comment.